src/crypto/tls/handshake_client_tls13.go

```diff
@@ -709,6 +709,30 @@ func certificateRequestInfo(certReq *certificateRequestMsgTLS13, vers uint16) *C
 	return cri
 }
 
+func getClientDelegatedCredential(cri *CertificateRequestInfo, cert *Certificate) (*DelegatedCredentialStruct, error) {
+	if len(cert.DelegatedCredentials) == 0 {
+		return nil, errors.New("No Delegated Credential found.")
+	}
+
+	if len(cert.DelegatedCredentials) == 1 {
+		// There's only one choice, so no point doing any work.
+		return &cert.DelegatedCredentials[0], nil
+	}
+
+	for _, dcStruct := range cert.DelegatedCredentials {
+		// If the client sent the signature_algorithms in the DC extension, ensure it supports
+		// schemes we can use with this delegated credential.
+		if len(cri.SignatureSchemesDC) > 0 {
+			if _, err := selectSignatureSchemeDC(VersionTLS13, dcStruct.DC, cri.SignatureSchemesDC); err == nil {
+				return &dcStruct, nil
+			}
+		}
+	}
+
+	// If nothing matches, return the first delegated credential.
+	return &cert.DelegatedCredentials[0], nil
```

Comment:
AFAICT, the client may use a DC whose algorithm isn't supported by the server. It seems to me that our options are either to fallback to signing with the leaf certificate or abort with "handshake_failure" if we don't have the signing key for the leaf certificate.

Similar comments for `getDelegatedCredential` below.

Reply:
This function will return a delegated credential, but at the handshake level, the check will be performed: https://github.com/cloudflare/go/blob/dc-api/src/crypto/tls/handshake_server_tls13.go#L522 https://github.com/cloudflare/go/blob/dc-api/src/crypto/tls/handshake_client_tls13.go#L806 

It errors because the usage of a delegated credential was advertised, a dc was found but it is invalid. It could be that we fallback to the certificate in that case as well, as the draft says:

```
If the extension is present, the server MAY send a delegated
credential; if the extension is not present, the server MUST NOT send
a delegated credential.
```

so it is always a MAY.